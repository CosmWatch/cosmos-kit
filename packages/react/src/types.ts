import {
  ChainRegistryInfo,
  ConnectedWallet,
  CosmosKitInitializeConfig,
  CosmosKitState,
  Wallet,
} from '@cosmos-kit/types'
import { ReactNode } from 'react'

export interface IWalletManagerContext extends CosmosKitState {
  // Function to begin the connection process. This will either display
  // the wallet picker modal or immediately attempt to connect to a wallet
  // depending on the props passed to WalletManagerProvider.
  connect: (wallet?: Wallet) => void
  // Function that disconnects from the connected wallet.
  disconnect: () => Promise<void>
  // If status is Connected.
  connected: boolean
}

export interface ModalClassNames {
  modalContent?: string
  modalOverlay?: string
  modalHeader?: string
  modalSubheader?: string
  modalCloseButton?: string
  walletList?: string
  wallet?: string
  walletImage?: string
  walletInfo?: string
  walletName?: string
  walletDescription?: string
  textContent?: string
}

export type UseWalletResponse = Partial<ConnectedWallet> &
  Pick<IWalletManagerContext, 'status' | 'connected' | 'error'>

export interface WalletManagerProviderProps
  extends CosmosKitInitializeConfig {
  children: ReactNode | ReactNode[]
  // Class names applied to various components for custom theming.
  classNames?: ModalClassNames
  // Custom close icon.
  closeIcon?: ReactNode
  // A custom loader to display in the modals, such as enabling the wallet.
  renderLoader?: () => ReactNode
  // chainInfo
  chainInfo: ChainRegistryInfo
}
