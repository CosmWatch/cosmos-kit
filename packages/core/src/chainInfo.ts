import { chainRegistryChainToKeplr } from '@chain-registry/keplr'
import { Chain } from '@chain-registry/types'
import { ChainRegistryInfo } from '@cosmos-kit/types'
import { ChainInfo } from '@keplr-wallet/types'

import { preferredChainInfo } from './preferred'

export const getKeplrChainInfo = async (
  chainName: string,
  info: ChainRegistryInfo
): Promise<ChainInfo> => {
  const chain: Chain | undefined = info.chains.find(
    (c) => c.chain_name === chainName
  )

  if (!chain) {
    throw new Error('cannot find chain: ' + chainName)
  }

  // you can add options as well to choose endpoints
  const config: ChainInfo = chainRegistryChainToKeplr(chain, info.assets, {
    getRestEndpoint: (chain) =>
      preferredChainInfo[chainName].rest ?? chain.apis?.rest[0]?.address,
    getRpcEndpoint: (chain) =>
      preferredChainInfo[chainName].rpc ?? chain.apis?.rpc[0]?.address,
  })

  return config
}
