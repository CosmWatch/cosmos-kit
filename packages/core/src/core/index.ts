export * from './api'
export * from './info'
export { addStateObservers, removeStateObserver, updateState } from './state'
