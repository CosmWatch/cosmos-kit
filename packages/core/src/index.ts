export * from './core'
export * from './walletInfo'
