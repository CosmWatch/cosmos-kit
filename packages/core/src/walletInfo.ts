import { SigningCosmWasmClientOptions } from '@cosmjs/cosmwasm-stargate'
import { SigningStargateClientOptions } from '@cosmjs/stargate'
import {
  ChainInfo,
  ConnectedWallet,
  Wallet,
  WalletAdapter,
} from '@cosmos-kit/types'

export const getConnectedWalletInfo = async <Client = unknown>(
  wallet: Wallet<Client>,
  adapter: WalletAdapter<Client>,
  client: Client,
  chainInfo: ChainInfo,
  signingCosmWasmClientOptions?: SigningCosmWasmClientOptions,
  signingStargateClientOptions?: SigningStargateClientOptions
): Promise<ConnectedWallet> => {
  await adapter.enableClient(client)

  // Parallelize for efficiency.
  const [{ name, address }, offlineSigner] = await Promise.all([
    // Get name and address.
    adapter.getNameAddress(),
    // Get offline signer.
    adapter.getOfflineSigner(),
  ])

  const [signingCosmWasmClient, signingStargateClient] = await Promise.all([
    // Get CosmWasm client.
    await (
      await import('@cosmjs/cosmwasm-stargate')
    ).SigningCosmWasmClient.connectWithSigner(
      adapter.getRpcEndpoint(),
      offlineSigner,
      signingCosmWasmClientOptions
    ),
    // Get Stargate client.
    await (
      await import('@cosmjs/stargate')
    ).SigningStargateClient.connectWithSigner(
      adapter.getRpcEndpoint(),
      offlineSigner,
      signingStargateClientOptions
    ),
  ])

  if (address === undefined) {
    throw new Error('Failed to retrieve wallet address.')
  }

  return {
    wallet,
    adapter,
    walletClient: client,
    chainInfo,
    offlineSigner,
    name,
    address,
    signingCosmWasmClient,
    signingStargateClient,
  }
}
