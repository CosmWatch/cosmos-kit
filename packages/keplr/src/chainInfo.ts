import { chainRegistryChainToKeplr } from '@chain-registry/keplr'
import { Chain } from '@chain-registry/types'
import { ChainRegistryInfo } from '@cosmos-kit/types'
import { ChainInfo as KeplrChainInfo } from '@keplr-wallet/types'

import { preferredChainInfo } from './preferred'

export const getKeplrChainInfo = (
  chainName: string,
  info: ChainRegistryInfo
): KeplrChainInfo => {
  const chain: Chain | undefined = info.chains.find(
    (c) => c.chain_name === chainName
  )

  if (!chain) {
    throw new Error('cannot find chain: ' + chainName)
  }

  // you can add options as well to choose endpoints
  const config: KeplrChainInfo = chainRegistryChainToKeplr(chain, info.assets, {
    getRestEndpoint: (chain) =>
      preferredChainInfo[chainName].rest ?? chain.apis?.rest[0]?.address,
    getRpcEndpoint: (chain) =>
      preferredChainInfo[chainName].rpc ?? chain.apis?.rpc[0]?.address,
  })

  return config
}
