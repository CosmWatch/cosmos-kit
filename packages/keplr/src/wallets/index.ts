export * from './keplr'
export * from './keplr-walletconnect'
