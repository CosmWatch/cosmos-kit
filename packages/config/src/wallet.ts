import { WalletInfo } from '@cosmos-kit/core';
import { ExtKeplrWallet, WCKeplrWallet } from '@cosmos-kit/keplr';
import { RiChromeFill, RiAppStoreFill } from "react-icons/ri";
import { GrFirefox } from "react-icons/gr";
import { FaAndroid } from "react-icons/fa";

export const walletInfos: WalletInfo[] = [
  {
    name: "keplr",
    active: false,
    logo: "https://pbs.twimg.com/profile_images/1498228570862219266/uctq7aeh_400x400.png",
    prettyName: "Keplr Wallet",
    wallet: new ExtKeplrWallet(),
    isQRCode: false,
    downloads: {
      desktop: [
        {
          browser: "chrome",
          icon: RiChromeFill,
          link: "https://chrome.google.com/webstore/detail/keplr/dmkamcknogkgcdfhhbddcghachkejeap?hl=en",
        },
        {
          browser: "firefox",
          icon: GrFirefox,
          link: "https://addons.mozilla.org/en-US/firefox/addon/keplr/",
        },
      ],
      tablet: [
        {
          os: "android",
          icon: FaAndroid,
          link: "https://addons.mozilla.org/en-US/firefox/addon/keplr/",
        },
        {
          os: "ios",
          icon: RiAppStoreFill,
          link: "https://addons.mozilla.org/en-US/firefox/addon/keplr/",
        },
      ],
      mobile: [
        {
          os: "android",
          icon: FaAndroid,
          link: "https://addons.mozilla.org/en-US/firefox/addon/keplr/",
        },
        {
          os: "ios",
          icon: RiAppStoreFill,
          link: "https://addons.mozilla.org/en-US/firefox/addon/keplr/",
        },
      ],
      default: "https://www.keplr.app/download",
    },
  },
  {
    name: 'wc-keplr',
    active: false,
    wallet: new WCKeplrWallet(),
    prettyName: 'Keplr Mobile',
    logo: "https://user-images.githubusercontent.com/545047/191616515-eee176d0-9e50-4325-9529-6c0019d5c71a.png",
    isQRCode: true
  }
];
