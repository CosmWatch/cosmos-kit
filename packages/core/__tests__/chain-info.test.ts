import { getChainInfo } from '../src/chainInfo';

it('works', async () => {
    const chain = await  getChainInfo('osmosis-1');
    expect(chain).toMatchSnapshot();
})