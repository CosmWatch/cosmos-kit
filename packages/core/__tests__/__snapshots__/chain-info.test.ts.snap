// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
Object {
  "bech32Config": Object {
    "bech32PrefixAccAddr": "osmo",
    "bech32PrefixAccPub": "osmopub",
    "bech32PrefixConsAddr": "osmovalcons",
    "bech32PrefixConsPub": "osmovalconspub",
    "bech32PrefixValAddr": "osmovaloper",
    "bech32PrefixValPub": "osmovaloperpub",
  },
  "bip44": Object {
    "coinType": 118,
  },
  "chainId": "osmosis-1",
  "chainName": "Osmosis",
  "currencies": Array [
    Object {
      "coinDecimals": 6,
      "coinDenom": "OSMO",
      "coinGeckoId": "osmosis",
      "coinImageUrl": "https://app.osmosis.zone/tokens/osmo.svg",
      "coinMinimalDenom": "uosmo",
    },
    Object {
      "coinDecimals": 6,
      "coinDenom": "ION",
      "coinGeckoId": "ion",
      "coinImageUrl": "https://app.osmosis.zone/tokens/ion.png",
      "coinMinimalDenom": "uion",
    },
  ],
  "features": Array [
    "stargate",
    "ibc-transfer",
    "no-legacy-stdTx",
    "ibc-go",
  ],
  "feeCurrencies": Array [
    Object {
      "coinDecimals": 6,
      "coinDenom": "OSMO",
      "coinGeckoId": "osmosis",
      "coinImageUrl": "https://app.osmosis.zone/tokens/osmo.svg",
      "coinMinimalDenom": "uosmo",
    },
  ],
  "gasPriceStep": Object {
    "average": 0,
    "high": 0.025,
    "low": 0,
  },
  "rest": "https://lcd-osmosis.keplr.app/",
  "rpc": "https://rpc-osmosis.keplr.app/",
  "stakeCurrency": Object {
    "coinDecimals": 6,
    "coinDenom": "OSMO",
    "coinGeckoId": "osmosis",
    "coinImageUrl": "https://app.osmosis.zone/tokens/osmo.svg",
    "coinMinimalDenom": "uosmo",
  },
}
`;
